src/QirRuntime/lib/QIR/allocationsTracker.cpp

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "allocationsTracker.hpp"
+
+#include "quantum__rt.hpp"
+
+using namespace std;
+
+namespace Microsoft
+{
+namespace Quantum
+{
+    void AllocationsTracker::OnAllocate(void* object)
+    {
+        auto inserted = this->allocatedObjects.insert(make_pair(object, 1));
+        if (inserted.second)
+        {
+            // first time we are allocating an object at this address, nothing to do
+        }
+        else
+        {
+            if (inserted.first->second > 0)
+            {
+                quantum__rt__fail(quantum__rt__string_create("Allocating an object over an existing object!"));
+            }
+            else
+            {
+                inserted.first->second = 1;
+            }
+        }
+    }
+
+    void AllocationsTracker::OnAddRef(void* object)
+    {
+        auto tracked = this->allocatedObjects.find(object);
+        if (tracked == this->allocatedObjects.end())
+        {
+            quantum__rt__fail(quantum__rt__string_create("Attempting to addref an object that isn't tracked!"));
+        }
+        else
+        {
+            if (tracked->second == 0)
+            {
+                quantum__rt__fail(quantum__rt__string_create("Attempting to ressurect a previously released object!"));
+            }
+            else
+            {
+                tracked->second += 1;
+            }
+        }
+    }
+
+    void AllocationsTracker::OnRelease(void* object)
+    {
+        auto tracked = this->allocatedObjects.find(object);
+        if (tracked == this->allocatedObjects.end())
+        {
+            quantum__rt__fail(quantum__rt__string_create("Attempting to release an object that isn't tracked!"));
+        }
+        else
+        {
+            if (tracked->second == 0)

```

Comment:
Minor:
As I understand the `this->allocatedObjects` is declared like this: 
std::unordered_map<void\*, **int**> allocatedObjects;
that is the _signed_ type (`int`) is used for the ref count (rather than an _unsigned_ type). Which means that theoretically the ref count can become below 0 (negative), either as a result of the memory corruption or in a hypothetical multithreading environment where 2 threads can call `AllocationsTracker::OnRelease()` in parallel for an object with ref count 1 (in which case the lines 67 - 70 below can get executed in parallel and in the line 69 the ref count can become `-1` for the slightly-later thread). If the ref count goes below 0 then during the subsequent `AllocationsTracker::OnRelease()` the line 63 will not catch that.
You may want to consider replacing the line 63 with `if (tracked->second <= 0)`
or to use the unsigned type (e.g. `unsigned int`) instead of `int` in the fragment 
std::unordered_map<void\*, **int**> allocatedObjects;

Reply:
QIR runtime is single threaded but it won't hurt to make the check broader here.